### DISCUSSION TITLE
Using OSXPhotos to synchronize folders and albums created in Apple Photos

### DISCUSSION THREAD

**Discussion by qkeddy** [RhetTbull/osxphotos#886] [Ideas]
With the release of macOS Ventura, there is a new feature called [iCloud Shared Photo Library](iCloud Shared Photo Library). It is convenient and does a good job of synchronizing videos and photos who are granted access as "Participants".  The one issue I have with this new feature is it doesn't synchronize the folders and albums. Could OSXPhotos help with this?

Here is the situation that I'm trying to solve:

1. My wife and I share an iCloud Shared Photo Library 
2. We have moved 100% of our photos on our film roll to be shared, and therefore we have 100% matching photos in our "Shared Library."
3. My wife does 100% of the organization of photos on her iPhone and MacBook. 
4. We would like to keep the same organization system using folders and albums in Photos.

A possible approach to solve this using OSXPhotos:

1. Using OSXPhotos, download in JSON format the folders and albums from my wife's Photos database.
2. The JSON file is saved to a shared iCloud file folder.
3. Using OSXPhotos, read from the shared iCloud file folder the JSON file and apply it to my Photos database.
4. Given that the photos are identical in the shared libraries, the folder & album organization should organize my Photos database.

Would it be possible to perform this synchronization with OSXPhotos? If so, what is the command line syntax? If not, could this be added as a feature to OSXPhotos?

**Comment by RhetTbull** [comment]
It is not currently possible to do this in osxphotos but the workflow you described *should* be relatively easy to implement. I really like this idea as a feature though and will open a new issue to add this. 

I envision something like a `sync-shared` command:

On computer 1:
`osxphotos sync-shared --export /path/to/shared/folder/computer1.db`

Then on computer 2:
`osxphotos sync-shared --import /path/to/shared/folder/computer1.db --album --title --description --keywords`

You could specify which metadata to sync.  (Does the shared library currently sync other metadata like titles and keywords?) 

I've already built a similar script for a user who did not use iCloud but wanted to sync metadata / albums between iPhone and Mac.  In that case, the difficult part was matching the photos as there wasn't a unique identifier to correlate the two libraries. In a regular iCloud library, each photo has a unique cloud identifier (`ZCLOUDMASTER.ZCLOUDMASTERGUID`).  For photos in a shared library, I'm not sure how Apple uniquely identifies the photos shared across two different libraries so that will have to be figured out. I have a machine running Ventura but have not used the shared library feature yet so don't have data to test on. 

edit: Rather than JSON, I'd probably use a sqlite database as the file format as it would be more performant and flexible (and still accessible via a variety of tools)

**Reply by RhetTbull** [reply]
@qkeddy As I don' t yet use the shared library feature, I don't have test data to reverse engineer how Apple is syncing the data between the two iCloud users.  If you're able to run the following commands on your computer and your wife's and send me the output (rturnbull+git@gmail.com) I can get started on the reverse engineering. I *think* the data is likely stored in the `ZSHARED` table (introduced in Big Sur and changed in Ventura).

On your computer:

`sqlite3 ~/Pictures/Photos\ Library.photoslibrary/database/Photos.sqlite ".headers on" "select * from ZSHARE" > computer1.txt`

On wife's computer:

`sqlite3 ~/Pictures/Photos\ Library.photoslibrary/database/Photos.sqlite ".headers on" "select * from ZSHARE" > computer2.txt`

Then send me `computer1.txt` and `computer2.txt`.  I don't think there will be any sensitive info in these files (based on the column names in the database) but you can take a look in TextEdit to verify before sending.

**Reply by qkeddy** [reply]
Thanks. I've sent to your email as requested.
